### PR TITLE
Use "" as the fallback from desc locales

### PR DESCRIPTION
**Pull request by CircuitSacul**
I decided to try my hand at this, since I hadn't worked with proc macros before.

This implementation uses the "" locale from the function as a fallback if there's no doc comment or desc attribute. This also means that ::create_command returns a result. The errors probably need some work.

I also added another test module for option descriptions, since that doesn't seem to be anywhere already.

I'm not sure if there's any other places where locale functions are used that could actually use this - I know that names cannot, since those are stored statically on the structs on expanding, at which point we don't have access to the HashMap of locales. Currently this handles command descriptions and option descriptions. Do choices have descriptions that would use this too?

I know you wanted to decide what implementation to use, which is fine. I wanted to do this anyways, even though there's a decent chance this won't be used. Since I actually succeeded I figured I'd open a draft pr so you could at least see it.

### CODE REVIEW

**Comment by baptiste0928** [comment]
Thanks for the PR! I accidentally turned off GitHub Actions checks for pull requests in a previous commit (fixed with #30). Can you merge `main` into your branch to have the checks run?

File: twilight-interactions-derive/src/command/model/create_command.rs [hunk 1]
```diff
@@ -42,8 +42,11 @@ pub fn impl_create_command(input: DeriveInput, fields: Option<FieldsNamed>) -> R
     };
     let name_localizations = localization_field(&attributes.name_localizations);
     let description = match &attributes.desc {
-        Some(desc) => desc.clone(),
-        None => parse_doc(&input.attrs, span)?,
+        Some(desc) => quote! { ::std::option::Option::Some((#desc).to_string()) },
+        None => match parse_doc(&input.attrs, span).ok() {
```
Comment (baptiste0928) [inline twilight-interactions-derive/src/command/model/create_command.rs:46]: We should change the `parse_doc` signature to  `Result<Option<String>, _>` instead, errors should not be ignored silently.

File: twilight-interactions-derive/src/command/model/create_command.rs [hunk 2]
```diff
@@ -97,9 +122,16 @@ fn field_option(field: &StructField) -> Result<TokenStream> {
     let name = field.attributes.name_default(field.ident.to_string());
     let name_localizations = localization_field(&field.attributes.name_localizations);
     let description = match &field.attributes.desc {
-        Some(desc) => desc.clone(),
-        None => parse_doc(&field.raw_attrs, field.span)?,
+        Some(desc) => quote! { ::std::option::Option::Some((#desc).to_string()) },
+        None => match parse_doc(&field.raw_attrs, field.span).ok() {
+            Some(doc) => quote! { ::std::option::Option::Some((#doc).to_string()) },
+            None => quote! { ::std::option::Option::None },
+        },
```
Comment (baptiste0928) [inline twilight-interactions-derive/src/command/model/create_command.rs:129]: Same as previous comments

File: twilight-interactions-derive/src/command/subcommand/create_command.rs [hunk 3]
```diff
@@ -31,8 +31,11 @@ pub fn impl_create_command(
     let name_localizations = localization_field(&attribute.name_localizations);
     let description_localizations = localization_field(&attribute.desc_localizations);
     let description = match attribute.desc {
-        Some(desc) => desc,
-        None => parse_doc(&input.attrs, span)?,
+        Some(desc) => quote! { ::std::option::Option::Some((#desc).to_string()) },
+        None => match parse_doc(&input.attrs, span).ok() {
+            Some(doc) => quote! { ::std::option::Option::Some((#doc).to_string()) },
+            None => quote! { ::std::option::Option::None },
+        },
```
Comment (baptiste0928) [inline twilight-interactions-derive/src/command/subcommand/create_command.rs:38]: Same as previous comments

**Review by baptiste0928 — COMMENTED** [review]
Thanks for your work on this! 

Regarding the feature design, I don't really like the fact that we lose compile-time validation. Maybe we can require the function to return a struct that explicitly defines the fallback instead:

```rs
fn locale_option() -> DescLocalizations {
    let locales = [("en", "en description")];
    
    // Using a constructor here to allow users to use different types:
    // locales => impl IntoIterator<Item = (ToString, ToString)>
    // fallback => impl ToString
    DescLocalizations::new(locales, "fallback description")
}
```

We can introduce a similar `NameLocalizations` type for consistency. What do you think of this alternative design?

**Comment by CircuitSacul** [comment]
> Thanks for your work on this!
> 
> Regarding the feature design, I don't really like the fact that we lose compile-time validation. Maybe we can require the function to return a struct that explicitly defines the fallback instead:
> 
> ```rust
> fn locale_option() -> DescLocalizations {
>     let locales = [("en", "en description")];
>     
>     // Using a constructor here to allow users to use different types:
>     // locales => impl IntoIterator<Item = (ToString, ToString)>
>     // fallback => impl ToString
>     DescLocalizations::new(locales, "fallback description")
> }
> ```
> 
> We can introduce a similar `NameLocalizations` type for consistency. What do you think of this alternative design?

I like it, the only concern I have is that this breaks all existing locale functions. This was already going to be a breaking change, so ig it's not that big a deal.

**Pull request closed without merging**